File: JY/3. Longest Substring Without Repeating Charaters.c
#define MAX(x,y) (x)>(y) ? (x) : (y)

int lengthOfLongestSubstring(char* s) {
    
    int s_length = 0;
    while (s[s_length] != '\0'){
        s_length++;
    }

    int *lastIndex = (int*)malloc(256*sizeof(int));
    for (int i =0 ; i <256 ; i++){
        lastIndex[i] = -1;
    }

    int start = 0 ;
    int max_length = 0;

    for (int end = 0; end < s_length ; end++){
        start = MAX(start, lastIndex[s[end]] +1 );
        max_length = MAX(max_length, end-start+1);
        lastIndex[s[end]] = end;
    }
    return max_length;
}

File: JY/3. Longest Substring Without Repeating Charaters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        ###sliding window 기법###
        distinct_alphabet = set()
        start = 0
        max_length = 0

        for i in range(len(s)):
            while s[i] in distinct_alphabet:
                distinct_alphabet.remove(s[start])
                start += 1
            distinct_alphabet.add(s[i])
            max_length = max(max_length, i-start +1)

        return max_length